Add Webhook support to sonarr

halkeye opened the pull request:
Add webhook support to sonarr

Add Form type url (type=url input field)
Add isValidUrl input type validation

halkeye commented:
its a start for #348

I'm working on a sonarr plugin for hubot - https://github.com/halkeye/hubot-sonarr

kayone commented on src/NzbDrone.Core/Notifications/Webhook/Webhook.cs:
```diff
@@ -0,0 +1,62 @@
+        {
+            get
+            {
+                return false;
```
any reasons why this was left out?

halkeye replied:
only because it only has a string message and no actual data? no real reason though.

halkeye replied:
and "CustomScript" skips it as well

kayone replied:
I think it would make sense to leave it in, for example if connecting it to slack or an IRC bot you might want to be notified of new grabs.

kayone replied:
custom script skips it generally because custom script is there to process the downloaded file, in case of grab there is nothing there to process.

halkeye replied:
How about we leave this one disabled, and I create a new patch that turns onGrab from being a string to something like DownloadMessage object?

Least then the webhook could post more than just a string?

I'm in no way against it, it just seems like a kinda hack.
I guess I could throw in the kinda hack until the next patch is done.

markus101 replied:
I think it makes sense to allow on grab, same as other "text" notifiers do.

kayone commented on src/NzbDrone.Common/Extensions/UrlExtensions.cs:
```diff
@@ -0,0 +1,30 @@
+                return false;
+
+            Uri uri;
+            if (!Uri.TryCreate(path, UriKind.RelativeOrAbsolute, out uri))
```
I don't think it makes sense to allow relative URLs, unless sonnar would want to send notifications to itself. :smile:

halkeye replied:
Very good point, fixing

halkeye commented:
@kayone I'll update this OnGrab if #760 gets accepted

markus101 commented on src/NzbDrone.Core/Notifications/Webhook/WebhookService.cs:
```diff
@@ -0,0 +1,92 @@
+
+        public void OnDownload(Series series, EpisodeFile episodeFile, WebhookSettings settings)
+        {
+            this.NotifyWebhook("Download", series, episodeFile, settings);
```
`this.` is not required (and we leave it out).

markus101 commented on src/NzbDrone.Core/Notifications/Webhook/WebhookSettings.cs:
```diff
@@ -0,0 +1,30 @@
+        private static readonly WebhookSettingsValidator Validator = new WebhookSettingsValidator();
+
+        [FieldDefinition(0, Label = "URL", Type = FieldType.Url)]
+        public String Url { get; set; }
```
nit: `string` instead of `String` (things are a bit all over with this one, but we've landed on using `string`).

halkeye replied:
Okay I've already shutdown for the night so I'll push a new patch in the
morning with all the stylistic changes in the morning.

If I can I'll throw in the text only on grab too but that might not be till
evening.
On Aug 23, 2015 11:09 PM, "Mark McDowall" <notifications@github.com> wrote:

> In src/NzbDrone.Core/Notifications/Webhook/WebhookSettings.cs
> <https://github.com/Sonarr/Sonarr/pull/759#discussion_r37721558>:
>
> > +namespace NzbDrone.Core.Notifications.Webhook
> > +{
> > +    public class WebhookSettingsValidator : AbstractValidator<WebhookSettings>
> > +    {
> > +        public WebhookSettingsValidator()
> > +        {
> > +            RuleFor(c => c.Url).IsValidUrl();
> > +        }
> > +    }
> > +
> > +    public class WebhookSettings : IProviderConfig
> > +    {
> > +        private static readonly WebhookSettingsValidator Validator = new WebhookSettingsValidator();
> > +
> > +        [FieldDefinition(0, Label = "URL", Type = FieldType.Url)]
> > +        public String Url { get; set; }
>
> nit: string instead of String (things are a bit all over with this one,
> but we've landed on using string).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Sonarr/Sonarr/pull/759/files#r37721558>.
>

halkeye commented:
I think i've fixed the outstanding issues @markus101 @kenyonj

kayone commented on src/NzbDrone.Core/Notifications/Webhook/WebhookService.cs:
```diff
@@ -0,0 +1,97 @@
+
+        public void OnDownload(Series series, EpisodeFile episodeFile, WebhookSettings settings)
+        {
+            NotifyWebhook(new { EventType = "Download", Series = series, EpisodeFile = episodeFile }, settings);
```
I would much prefer us having a separate data model we use for webhook payload instead of relying on `Series`, `EpisodeFile` which are considered internal contracts. that way if we add/remove/rename stuff from our internal models the webhook implementations are protected by the mapping.

it would be great if you could create a new class, `WebHookSeries`, `WebHookEpisode` and add whatever fields from `Series`, `Episodes` that you think makes sense to an external application. we want to be selective instead of copying the whole thing over since we can add stuff later but once something is added it can't be removed since it would be considered a breaking change.

kayone commented:
sorry if we are being a bit picky, but features like this are extremely hard to change once they are merged since they would be considered part of our API and changing them could break things, that's why it's very important to get them done as well as we possibly can on the initial run.

halkeye commented:
Its all good.

Should I just base it on the same fields that customscript uses?

kayone commented:
yes, that would be great, but break the fields into Series, Episode objects,

@markus101 anything we would change about custom scripts here (thinking of it as a second change)?

markus101 commented:
Nothing comes to mind, webhooks should be a bit cleaner since we can serialize to JSON though :+1:

halkeye commented:
Who creates the new object? Webhook.cs? Or WebhookService.cs?

kayone commented:
I think it would be the service,

halkeye commented:
I wasn't sure if a helper function should be created, or it made sense to put in the constructor. I can code in C# but don't use it on a day to day basis.

halkeye commented:
@kayone @markus101 Any more feedback?
(not trying to be a pest, but I like to finish coding before i forget too much about what I was working on)

kayone commented on src/NzbDrone.Common/Extensions/UrlExtensions.cs:
```diff
@@ -0,0 +1,30 @@
+    {
+        public static bool IsValidUrl(this string path)
+        {
+            if (string.IsNullOrWhiteSpace(path))
```
please add `{}`

halkeye replied:
just a code style thing? I'll update my patch tongiht

kayone replied:
yeah, mainly, it's more explicit about where the if boundaries start and end, leaves less room for mistakes.

kayone commented on src/NzbDrone.Core/Notifications/Webhook/Webhook.cs:
```diff
@@ -0,0 +1,55 @@
+
+        public override string Link
+        {
+            get { return "https://github.com/Sonarr/Sonarr/wiki/Webhook"; }
```
can you add a wiki page with sample requests that are sent by the webhook, example of the body and headers would be great, maybe mention that it's limited to `POST` and `JSON` for now,

halkeye replied:
Updated

kayone commented on src/NzbDrone.Core/Notifications/Webhook/WebhookSeries.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    internal class WebhookSeries
+    {
+
```
extra blank space,

kayone commented on src/NzbDrone.Core/Notifications/Webhook/WebhookPayload.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    internal class WebhookPayload
+    {
+        public WebhookEpisodeFile EpisodeFile { get; internal set; }
```
can you move this to the end,

kayone commented on src/NzbDrone.Core/Notifications/Webhook/WebhookSeries.cs:
```diff
@@ -0,0 +1,21 @@
+    internal class WebhookSeries
+    {
+
+        public string Id { get; set; }
```
any reason you didn't use it as `int`?

halkeye replied:
Nope, nice catch, was probably because I was autogenerating them and it still had the .ToString().

I'll update

kayone commented on src/NzbDrone.Core/Notifications/Webhook/WebhookSeries.cs:
```diff
@@ -0,0 +1,21 @@
+        public string Id { get; set; }
+        public string Title { get; set; }
+        public string Path { get; set; }
+        public string TvdbId { get; set; }
```
same here,

kayone commented on src/NzbDrone.Core/Notifications/Webhook/WebhookService.cs:
```diff
@@ -0,0 +1,101 @@
+            try {
+                var client = RestClientFactory.BuildClient(settings.Url);
+                var request = new RestRequest(Method.POST);
+                request.RequestFormat = DataFormat.Json;
```
I'm thinking we might want to make the `ContentType` and `Method` configurable as part of settings, but not something we need right now,

halkeye replied:
I don't think GET supports content type right? The rest would. If I did GET I'd have to append everything onto the URL, which would get pretty messy unless it just appended a giant json blob?

Could easily support PUT & PATCH as long as restsharp supports it

kayone replied:
I don't think `PATCH` makes sense either, it's too semantically wrong. I would say `PUT` and `POST` are the ones that make sense.

kayone commented:
once we merge https://github.com/Sonarr/Sonarr/pull/760 this should be able to get in with no issues.

Thanks for the great patch!

kayone commented:
I merged your other branch, you can rebase from develop to pickup the changes,

halkeye commented:
@kayone 
Rebased from develop. Added Method Selection and a few other cleanups

Why does EpisodeFile have a list of episodes? Is it for like a seasonpack? I think its safe for removal in this stuff?

Added wiki page 
https://github.com/Sonarr/Sonarr/wiki/Webhook

markus101 commented:
> Why does EpisodeFile have a list of episodes? Is it for like a seasonpack? I think its safe for removal in this stuff?

Files with more than one episode, double episodes usually, not ever likely to see a full season in a single file (which would be rejected on import anyways).

Currently each notification is for a single episode so there is no need to have a list of them, though that will likely change later (reduce notifications when importing a season pack). Personally I'd expect the episode to be a separate subobject in the JSON.

markus101 commented on src/NzbDrone.Core/Notifications/Webhook/WebhookEpisodeFile.cs:
```diff
@@ -0,0 +1,33 @@
+            RelativePath = episodeFile.RelativePath.ToString();
+            FullPath = Path.Combine(series.Path, episodeFile.RelativePath).ToString();
+            SeasonNumber = episodeFile.SeasonNumber;
+            Episodes = episodeFile.Episodes.Value.ConvertAll(x => new WebhookEpisode(x));
```
This seems redundant when Episode is already on the main object.

markus101 commented on src/NzbDrone.Core/Notifications/Webhook/WebhookEpisode.cs:
```diff
@@ -0,0 +1,20 @@
+
+        public string AirDate { get; internal set; }
+        public DateTime? AirDateUtc { get; internal set; }
+        public int EpisodeNumber { get; internal set; }
```
How about the episode title?

halkeye replied:
Updated

kayone replied:
still missing the title?

markus101 commented on src/NzbDrone.Core/Notifications/Webhook/WebhookEpisodeFile.cs:
```diff
@@ -0,0 +1,33 @@
+        public int SeasonNumber { get; set; }
+        public List<WebhookEpisode> Episodes { get; set; }
+        public string Quality { get; set; }
+        public string QualityVersion { get; set; }
```
Why convert the int to a string?

halkeye replied:
Good catch. I missed one from my copy and paste job from customscript

halkeye commented:
> Currently each notification is for a single episode so there is no need to have a list of them, though that will likely change later (reduce notifications when importing a season pack). Personally I'd expect the episode to be a separate subobject in the JSON.

Me too, I'll remove the list from the EpisodeFile item

kayone commented on src/NzbDrone.Core/Notifications/Webhook/WebhookPayload.cs:
```diff
@@ -0,0 +1,10 @@
+        public string EventType { get; internal set; }
+        public WebhookSeries Series { get; internal set; }
+        public WebhookEpisodeFile EpisodeFile { get; internal set; }
+        public WebhookEpisode Episode { get; internal set; }
```
can we make this a list, that way in the future when we add support for bulk notifications (you get notification for a season pack) we don't have to break the API.

halkeye replied:
2 lists:
List<EpisodeFile> + List<Episode>
or 1 list:
List<Episode,EpisodeFile>

markus101 replied:
If we make episodes a list we'd need to make episode files a list as well since they could both be plural.

kayone replied:
do we really need two separate items? couldn't we just have a list of episodes that contains the file info as well, right now it appears that random fields are spread across the two models,

kayone replied:
just to clarify, I'm not saying it's random, I'm saying to a typical user of this API it would appear that it's random.

halkeye replied:
> couldn't we just have a list of episodes that contains the file info as well

yea, that seems a lot cleaner to me. 

I need to figure out how to get restsharp to exclude nulls, but ongrab won't have any file information which kinda makes sense to me? I think i can massage the data to get the quality and release/scene group

kayone replied:
so here is what I'm thinkings, we break what is currently `EpisodeFile` into two parts, the release info and file info,

we keep the release info separate as single item (since even a season pack would only have 1 release info), release info would be `sceneName`, `quality`, `releaseGroup` etc.

and we embed the file info into each episode, some will have duplicated info but if a consumer cares they can match items using the file path.

halkeye replied:
so like:

```json
Episode: {
Airdate: blah,
Airdateutc
EpisodeNumber
SeasonNumber,
Title
},
ReleaseInfo: {
sceneName,
quality,
releaseGroup
}
```

I think sometimes, especially with torrents, season packs can have different scene names and quality?

kayone replied:
if I remember correctly, as far as sonarr is concerned scene name is the name of the torrent/nzb, we don't actually use the internal filename of episodes inside of a pack, so `SceneName` should be unique. as for quality, that is a good point,

we could always move the quality from `ReleaseInfo` to `Episode`

markus101 replied:
Scene name is usually the release (folder) name, but in a season pack it could be the file name. Its very unlikely that a season pack would have multiple qualities though.

kayone replied:
@markus101 so what do you think, include everything in the `episode` object? is there a point in having release and episdoe separated?

halkeye replied:
So i'll have time to work on it again tonight. I'm going to move forward with making a list of episodes and throwing in the data in there.

kayone commented on src/NzbDrone.Core/Notifications/Webhook/WebhookEpisodeFile.cs:
```diff
@@ -0,0 +1,31 @@
+            Quality = episodeFile.Quality.Quality.Name;
+            QualityVersion = episodeFile.Quality.Revision.Version;
+            ReleaseGroup = episodeFile.ReleaseGroup ?? string.Empty;
+            SceneName = episodeFile.SceneName ?? string.Empty;
```
I'm a bit torn on this, @markus101 what do you think? should we send `''`back when these are empty or keep them `null`?

markus101 replied:
I think null makes the most sense.

halkeye commented:
@kayone @markus101 Updated the code as per suggestions (I think, this pull request is getting messy).

I've also updated the wiki page with the the sample outputs. https://github.com/Sonarr/Sonarr/wiki/Webhook

markus101 commented on src/NzbDrone.Core/Notifications/Webhook/WebhookService.cs:
```diff
@@ -0,0 +1,113 @@
+
+        public void OnDownload(Series series, EpisodeFile episodeFile, WebhookSettings settings)
+        {
+            WebhookPayload payload = new WebhookPayload
```
`var` instead of `WebhookPayload` would be more consistent with what we normally use.

halkeye replied:
@markus101 replaced WebhookPayload with var

markus101 commented on src/NzbDrone.Core/Notifications/Webhook/WebhookService.cs:
```diff
@@ -0,0 +1,113 @@
+            NotifyWebhook(payload, settings);
+        }
+
+        public void NotifyWebhook(object body, WebhookSettings settings)
```
`public void NotifyWebhook(WebhookPayload payload, WebhookSettings settings)`

halkeye replied:
How should I handle test then? Make a fake series? or just null both?

kayone replied:
I think that would make more sense, the other option which I think would be ok is to add a default constructor to WebhookPayload` and all it's subclasses, that way you can create a custom one rather than always having to create them using `Series` etc.

kayone commented on src/NzbDrone.Core/Notifications/Webhook/WebhookPayload.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    internal class WebhookPayload
+    {
+        public string EventType { get; internal set; }
```
I think it's ok to make these `public` we generally try to avoid using `internal`, 

internal is normally used for libraries where you share your code and you don't want 3rd parties to access certains parts of it, which isn't the case with us.

halkeye replied:
it was autogenerated :) Just make the set normal then?

kayone replied:
yeah, I would say make everything that is internal either private or public.

halkeye commented:
@kayone okay, everything is using the webhook object. Also no more internal

kayone commented on src/NzbDrone.Core/Notifications/Webhook/WebhookService.cs:
```diff
@@ -0,0 +1,129 @@
+    public class WebhookService : IWebhookService
+    {
+        private readonly IProcessProvider _processProvider;
+        private readonly Logger _logger;
```
I don't think these two are being used.

halkeye replied:
Yep yep. Removed

kayone commented:
@halkeye Again, thanks for contributing, sorry for making you run around a bit :smile: 

@markus101 any blockers before we merge this in?

halkeye commented:
No problem, Its a feature I want, so I want it to be right.

Looks like my apt's power is flickering, so I might not be able to do much more today

kayone commented on src/NzbDrone.Core/Notifications/Webhook/WebhookService.cs:
```diff
@@ -0,0 +1,124 @@
+
+    public class WebhookService : IWebhookService
+    {
+        public WebhookService(IProcessProvider processProvider, Logger logger)
```
you don't actually need the constructor anymore since it doesn't actually do anything. You can just kill it.

kayone commented:
It is pretty windy out there :D

halkeye commented:
@kayone remove constructor

Its kinda fun watching the various lose stuff flying around out there

markus101 commented:
Nope, it was all good from my end. Power is out though, ugh.

kayone commented:
@halkeye I can merge this in once you squash and rebase it.

halkeye commented:
yay awesome~

Squashed!

I kept all the commit messages as one, for future reference, do you want me to delete all the extra commit messages?

halkeye commented:
Can I add https://github.com/halkeye/hubot-sonarr/ to https://github.com/Sonarr/Sonarr/wiki/Webhook as an example app that supports sonarr webhooks?

markus101 commented:
Please do, hopefully it can help someone else get started.

kayone commented:
@halkeye we generally don't really worry about different commit messages, if the changes are isolated we keep the commits separated, but a feature like this fits well with a single commit. 

also, please make the commit message something that starts with "New:" like "New: Added support for webhook notifications" that way it'll show up in the change log.

halkeye commented:
@kayone I've updated the commit message

kayone commented:
Merged, Thank you!
